mllib/src/main/scala/org/apache/spark/mllib/stat/test/KSTest.scala

```diff
@@ -0,0 +1,203 @@
+ * cumulative distribution value. We know the latter to be correct, while the former will be off by
+ * a constant (how large the constant is depends on how many values precede it in other partitions).
+ * However, given that this constant simply shifts the ECDF upwards, but doesn't change its shape,
+ * and furthermore, that constant is the same within a given partition, we can pick 2 values
+ * in each partition that can potentially resolve to the largest global distance. Namely, we
+ * pick the minimum distance and the maximum distance. Additionally, we keep track of how many
+ * elements are in each partition. Once these three values have been returned for every partition,
+ * we can collect and operate locally. Locally, we can now adjust each distance by the appropriate
+ * constant (the cumulative sum of # of elements in the prior partitions divided by the data set
+ * size). Finally, we take the maximum absolute value, and this is the statistic.
+ */
+private[stat] object KSTest extends Logging {
+
+  // Null hypothesis for the type of KS test to be included in the result.
+  object NullHypothesis extends Enumeration {
+    type NullHypothesis = Value
+    val oneSampleTwoSided = Value("Sample follows theoretical distribution")
+  }
+
+  /**
+   * Runs a KS test for 1 set of sample data, comparing it to a theoretical distribution
+   * @param data `RDD[Double]` data on which to run test
+   * @param cdf `Double => Double` function to calculate the theoretical CDF
+   * @return KSTestResult summarizing the test results (pval, statistic, and null hypothesis)
+   */
+  def testOneSample(data: RDD[Double], cdf: Double => Double): KSTestResult = {
+    val n = data.count().toDouble
+    val localData = data.sortBy(x => x).mapPartitions { part =>
+      val partDiffs = oneSampleDifferences(part, n, cdf) // local distances
+      searchOneSampleCandidates(partDiffs) // candidates: local extrema
+    }.collect()
+    val ksStat = searchOneSampleStatistic(localData, n) // result: global extreme
+    evalOneSampleP(ksStat, n.toLong)
+  }
+
+  /**
+   * Runs a KS test for 1 set of sample data, comparing it to a theoretical distribution
+   * @param data `RDD[Double]` data on which to run test
+   * @param createDist `Unit => RealDistribution` function to create a theoretical distribution
+   * @return KSTestResult summarizing the test results (pval, statistic, and null hypothesis)
+   */
+  def testOneSample(data: RDD[Double], createDist: () => RealDistribution): KSTestResult = {
+    val n = data.count().toDouble
+    val localData = data.sortBy(x => x).mapPartitions { part =>
+      val partDiffs = oneSampleDifferences(part, n, createDist) // local distances
+      searchOneSampleCandidates(partDiffs) // candidates: local extrema
+    }.collect()
+    val ksStat = searchOneSampleStatistic(localData, n) // result: global extreme
+    evalOneSampleP(ksStat, n.toLong)
```

Comment:
@mengxr I've refactored the code, accounting that RealDistribution in math3 3.4.1 is serializable (which Sean had already pointed out to me). So that we now indeed use the testOneSample(..., cdf) and avoid duplication.